clang/lib/CIR/Dialect/Transforms/LoweringPrepareItaniumCXXABI.cpp

```diff
@@ -115,3 +110,9 @@ LoweringPrepareItaniumCXXABI::lowerDynamicCast(CIRBaseBuilderTy &builder,
           })
       .getResult();
 }
+
+mlir::Value LoweringPrepareItaniumCXXABI::lowerVAArg(CIRBaseBuilderTy &builder,
+                                                     mlir::cir::VAArgOp op) {
+  // TODO: implement va_arg for more generic Itanium ABI.
+  return mlir::Value();
```

Comment:
My original thought was for ABIs such as arm32 or other ABIs where var_arg lowering has much simpler logic, like is just loading from stack, it could be done during LLVMLowering, see https://godbolt.org/z/ccoKzEoe7
thus here we simply pass and do nothing
But now I agree with your more cautious approach, just let it fail here and force whoever implements LLVMLowering of a ABI to revisit this code.

Reply:
Not sure I understand your point, regardless if other paths already cover NYI we should always be explicit and also use `llvm_unreachable("NYI");` here.

Reply:
> Not sure I understand your point, regardless if other paths already cover NYI we should always be explicit and also use `llvm_unreachable("NYI");` here.

No problem put llvm_unreachable("NYI") here but then we need to change test variadics.c which expects for x86-64, Cir gen to be successful and preserve cir.var_arg.  But of course we can change the test and let it test --mlir-print-ir-before=cir-lowering-prepare?

Reply:
> No problem put llvm_unreachable("NYI") here but then we need to change test variadics.c which expects for x86-64, Cir gen to be successful and preserve cir.var_arg. But of course we can change the test and let it test --mlir-print-ir-before=cir-lowering-prepare?

Right now (before your PR), `clang/test/CIR/CodeGen/variadics.c` has 3 run lines:
- Generic for c (CHECK)
- x86 for c++ (CHECK)
- arm for c++ (CHECK)

The only changes you should do here are:
- Generic for c -> Make it x86 or ARM before lowering prepare for C, just one of the arches is enough (CHECK)
- x86 before lowering prepare for c++ (CHECK)
- arm before lowering prepare for c++ (CHECK)

No need to add new checks. Note that "(CHECK)" above means they should all use the same check prefix. There's no need to introduce a new ARM64_CHECK here, since they are currently all equal before lowering prepare.

All the "after-lowering prepare" logic is already tested in the other files you added: `clang/test/CIR/CodeGen/var-arg.c` and `clang/test/CIR/CodeGen/var-arg-float.c`.